File: packages/knowledge-engine/src/terminology/term.ts
import type { ContentLanguage } from '../topics/topic';

/**
 * An entry in the Terminology Dictionary (`content/terminology/`, `10` § Terminology Dictionary).
 *
 * The rule the whole file exists to express: **the term is preserved, the explanation is localized.**
 *
 *     Connection Pooling
 *     Açılan veritabanı bağlantılarının kapatılmak yerine bir havuzda tutulup yeniden kullanılmasıdır.
 *
 * The Turkish sentence is Turkish. `Connection Pooling` is still `Connection Pooling`, because that is
 * what the reader will type into a search box, read in a stack trace, and see in a job advert. A topic
 * that teaches them "bağlantı havuzlama" has taught them a word nobody else uses.
 */
export interface Term {
  /** The canonical spelling. This exact string must survive translation. */
  term: string;

  /**
   * Other spellings that count as the term: `DI` for `Dependency Injection`, `GC` for
   * `Garbage Collector`. Present in the source and an alias present in the translation is not a
   * violation — an abbreviation is still the term.
   */
  aliases: string[];

  /**
   * Translations that are WRONG, listed by name.
   *
   * This is the half that catches the failure that prompted the rule: an AI translator writing
   * "havuzdakiler" instead of "Connection Pool'daki bağlantılar". The preservation rule alone would
   * catch it only if the term vanished entirely — but a translator usually keeps ONE occurrence and
   * paraphrases the other five. Naming the paraphrase makes every occurrence checkable.
   */
  forbiddenTranslations: string[];

  /** The localized explanation. The only part of a term that may be translated. */
  explanations: Partial<Record<ContentLanguage, string>>;
}

/**
 * Does `text` contain `term`, allowing for Turkish inflection?
 *
 * Turkish agglutinates: `Connection Pool` becomes `Connection Pool'daki`, `Middleware` becomes
 * `Middleware'in`, and a naive whole-word match would call both of those a missing term. The suffix
 * belongs to the language; the term is intact underneath it, and that is what we are checking.
 *
 * Matching is therefore: the term, at a word boundary, optionally followed by an apostrophe and a
 * suffix. Case-insensitive, because a term at the start of a sentence is still the term.
 *
 * What this deliberately does NOT do is match a term embedded in a longer word — `Pooling` inside
 * `Poolinglemek` is not the term, it is somebody inventing a verb.
 */
export function containsTerm(text: string, term: string): boolean {
  const escaped = term.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  // Word boundary, the term, then either an apostrophe + suffix (Turkish inflection), or a boundary.
  // \p{L} keeps this correct for Turkish letters, which \w does not cover.
  const pattern = new RegExp(`(?<![\\p{L}\\d])${escaped}(?:['’]\\p{L}+)?(?![\\p{L}\\d])`, 'iu');

  return pattern.test(text);
}

/** Every spelling that counts as this term: the canonical one, plus its aliases. */
export function spellingsOf(entry: Term): string[] {
  return [entry.term, ...entry.aliases];
}

File: apps/api/src/WhyStack.Domain/Content/Term.cs
namespace WhyStack.Domain.Content;

/// <summary>
/// An approved technical term (`10` § Terminology Dictionary; `content/terminology/` under ADR-0018).
/// </summary>
/// <remarks>
/// The rule this table exists for: <b>the term is preserved, the explanation is localized.</b>
///
///     Connection Pooling
///     Açılan veritabanı bağlantılarının kapatılmak yerine bir havuzda tutulup yeniden kullanılmasıdır.
///
/// The Turkish sentence is Turkish. `Connection Pooling` is still `Connection Pooling`, because that is what
/// the reader will type into a search box, read in a stack trace, and see in a job advert. A topic that
/// teaches them "bağlantı havuzlama" has taught them a word nobody else uses — and has made them worse at
/// finding the answer next time.
///
/// It moved here from a YAML file with the rest of the content (ADR-0020). It is content: an editor adds a
/// term the same way they add a topic, and the validator picks it up on the next save.
/// </remarks>
public class Term
{
    public Guid Id { get; init; }

    /// <summary>The canonical spelling. This exact string must survive translation.</summary>
    public required string Text { get; set; }

    /// <summary>
    /// Other spellings that count as the term — `DI`, `GC`, `CLR`. Semicolon-separated.
    /// </summary>
    /// <remarks>
    /// A child table would be correct and would buy nothing: nothing queries an alias, nothing indexes one,
    /// and the whole dictionary is loaded into memory to validate a single save. A join per term, to read a
    /// list of two strings, is a join for the schema diagram's benefit.
    /// </remarks>
    public string Aliases { get; set; } = string.Empty;

    /// <summary>
    /// Translations that are WRONG, listed by name. Semicolon-separated.
    /// </summary>
    /// <remarks>
    /// The half that catches the realistic failure. A translator — human or model — rarely drops a term
    /// entirely; it keeps it in the heading, where it looks diligent, and paraphrases it for the next five
    /// paragraphs. "The term survived" would pass that. Naming the paraphrase makes every occurrence
    /// checkable.
    /// </remarks>
    public string ForbiddenTranslations { get; set; } = string.Empty;

    public ICollection<TermExplanation> Explanations { get; init; } = [];
}

/// <summary>The term, explained in one language. The explanation is the only part that may be translated.</summary>
public class TermExplanation
{
    public Guid Id { get; init; }

    public required Guid TermId { get; init; }

    public required string LanguageCode { get; init; }

    public required string Text { get; set; }
}
